doc/rst/platforms/comm-layers/ugni.rst

```diff
@@ -0,0 +1,228 @@
+   support.  For example, ``craype-hugepages16M`` supports 16 MiB
+   hugepages.  It does not matter which ``craype-hugepages`` module you
+   have loaded when you build your program.  Any of them will do.  Which
+   one you have loaded when you run a program does matter, however.  For
+   general use, the Chapel group recommends the ``craype-hugepages16M``
+   module.  You can read on for more information about hugepage modules
+   if you would like, but the recommended ``craype-hugepages16M`` module
+   will probably give you satisfactory results.
+
+   The Cray network interface chips (NICs) can only address memory that
+   has been registered with them. In practical terms, the Aries(TM) NIC
+   on Cray XC systems is not limited as to how much memory it can
+   register.  However, it does have an on-board cache of 512 registered
+   page table entries, and registering more than this can cause reduced
+   performance if the program's memory reference pattern causes refills
+   in this cache.  We have seen up to a 15% reduction from typical
+   nightly XC-16 performance in an ra-rmo run using hugepages small
+   enough that every reference should have missed in this cache.
+   Covering an entire 128 GiB XC compute node with only 512 hugepages
+   will require at least the ``craype-hugepages256M`` module's 256 MiB
+   hugepages.
+
+   Offsetting this, using larger hugepages may reduce performance because
+   it can result in poorer NUMA affinity.  With the ugni communication
+   layer, arrays larger than 2 hugepages are allocated separately from the
+   heap, which improves NUMA affinity.  An obvious side effect of using
+   larger hugepages is that an array has to be larger to qualify.  Thus,
+   achieving the best performance for any given program may require
+   striking a balance between using larger hugepages to reduce NIC page
+   table cache refills and using smaller ones to improve NUMA locality.
+
+   Note that when hugepages are used with the ugni comm layer, tasking
+   layers cannot use guard pages for stack overflow detection.  Qthreads
+   tasking cannot detect stack overflow except by means of guard pages,
+   so if ugni communications is combined with qthreads tasking and a
+   hugepage module is loaded, stack overflow detection is unavailable.
+
+
+Network Atomics
+_______________
+
+The Aries networks on Cray XC series systems support remote atomic
+memory operations (AMOs).  When the ``CHPL_NETWORK_ATOMICS`` environment
+variable is set to ``ugni``, the following operations on remote atomics
+are done using the network::
+
+    32- and 64-bit signed and unsigned integer types:
+    32- and 64-bit real types:
+      read()
+      write()
+      exchange()
+      compareAndSwap()
+      add(), fetchAdd()
+      sub(), fetchSub()
+
+    32- and 64-bit signed and unsigned integer types:
+      or(),  fetchOr()
+      and(), fetchAnd()
+      xor(), fetchXor()
```

Comment:
There's a part of me that wants these to be bulleted lists instead of the odd code block things they turn into.  But that's a pre-existing issue and really feels like small potatoes compared to getting this change in